keanu-project/src/main/java/io/improbable/keanu/distributions/discrete/Categorical.java

```diff
@@ -18,44 +20,51 @@ private Categorical(Map<T, DoubleTensor> selectableValues) {
         this.selectableValues = selectableValues;
     }
 
-    public T sample(int[] shape, KeanuRandom random) {
-        double sumOfProbabilities = getSumOfProbabilities();
-        double p = random.nextDouble();
-        double sum = 0;
-
-        if (sumOfProbabilities == 0.0) {
+    public GenericTensor<T> sample(int[] shape, KeanuRandom random) {
+        DoubleTensor sumOfProbabilities = getSumOfProbabilities(shape);
+        if (!sumOfProbabilities.lessThanOrEqual(0.).allFalse()) {
             throw new IllegalArgumentException("Cannot sample from a zero probability setup.");
         }
 
-        T value = null;
+        double p = random.nextDouble();
+        DoubleTensor sum = DoubleTensor.zeros(shape);
+        GenericTensor<T> value = new GenericTensor<>(shape);
+
         for (Map.Entry<T, DoubleTensor> entry : selectableValues.entrySet()) {
-            sum += entry.getValue().scalar() / sumOfProbabilities;
-            if (p < sum) {
-                value = entry.getKey();
+            sum.plusInPlace(entry.getValue().div(sumOfProbabilities));
+
+            BooleanTensor mask = sum.greaterThan(p);
+            value.setWithMaskInPlace(mask.toDoubleMask(), entry.getKey());
```

Comment:
Thoughts as a reader: Why does this work? Is there a consequence of setting this at every iteration, as opposed to once in the original version?